File: MORE/LAADS/MOD/MOD04_3K_A2022022_2355_061_2022024190628_hdf.m
%
%  This example code illustrates how to access and visualize MOD04_3K L2 file
% in MATLAB. 
%
%  If you have any questions, suggestions, comments on this example, please 
% use the HDF-EOS Forum (http://hdfeos.org/forums). 
% 
%  If you would like to see an  example of any other NASA HDF/HDF-EOS data 
% product that is not listed in the HDF-EOS Comprehensive Examples page 
% (http://hdfeos.org/zoo), feel free to contact us at eoshelp@hdfgroup.org 
% or post it at the HDF-EOS Forum (http://hdfeos.org/forums).
%                                   
% Usage:save this script and run (without .m at the end)
%                                   
% $matlab -nosplash -nodesktop -r MOD04_3K_A2022022_2355_061_2022024190628_hdf
%
% Tested under: MATLAB R2021a
% Last updated: 2022-01-27
%

import matlab.io.hdfeos.*
import matlab.io.hdf4.*
                                   
FILE_NAME='MOD04_3K.A2022022.2355.061.2022024190628.hdf';
SWATH_NAME='mod04';
% Set data field to read.                                   
% DATAFIELD_NAME = 'Angstrom_Exponent_1_Ocean';
DATAFIELD_NAME ='Optical_Depth_Land_And_Ocean';

file_id = sw.open(FILE_NAME, 'rdonly');
swath_id = sw.attach(file_id, SWATH_NAME);
data_i16 = sw.readField(swath_id, DATAFIELD_NAME, [], [], []);
                                   
% Read lat and lon.
lon = sw.readField(swath_id, 'Longitude', [], [], []);
lat = sw.readField(swath_id, 'Latitude', [], [], []);

% Detach Swath object.
sw.detach(swath_id);
sw.close(file_id);

lat=double(lat);
lon=double(lon);

% Read attributes from the data field.
SD_id = sd.start(FILE_NAME, 'rdonly');
sds_index = sd.nameToIndex(SD_id, DATAFIELD_NAME);
sds_id = sd.select(SD_id, sds_index);

% Read _FillValue from data field.
fillvalue_index = sd.findAttr(sds_id, '_FillValue');
fillvalue = sd.readAttr(sds_id, fillvalue_index);

% Get the long name from data field.
long_name_index = sd.findAttr(sds_id, 'long_name');
long_name = sd.readAttr(sds_id, long_name_index);

% Read units from the data field.
units_index = sd.findAttr(sds_id, 'units');
units = sd.readAttr(sds_id, units_index);

% Read scale_factor from the data field.
scale_index = sd.findAttr(sds_id, 'scale_factor');
scale = sd.readAttr(sds_id, scale_index);
scale = double(scale);

% Read add_offset from the data field.
offset_index = sd.findAttr(sds_id, 'add_offset');
offset = sd.readAttr(sds_id, offset_index);
offset = double(offset);

% Terminate access to the corresponding data set.
sd.endAccess(sds_id);

% Close the file.
sd.close(SD_id);

data = double(data_i16);

% Replace fill value with NaN.
data(data_i16==fillvalue) = NaN;

% Apply scale and offset.
data = scale*(data-offset);

% Map to plot the results.
latlim=double([floor(min(min(lat))),ceil(max(max(lat)))]);
lonlim=double([floor(min(min(lon))),ceil(max(max(lon)))]);


% Plot wind speed.
g = figure('Name', FILE_NAME, 'visible', 'off');
axesm('MapProjection','eqdcylin',...
      'Frame','on','Grid','on', ...
      'MeridianLabel','on','ParallelLabel','on','MLabelParallel', ...
      'south', 'MapLatLimit',latlim,'MapLonLimit',lonlim)
coast = load('coastlines.mat');

scatterm(lat(:), lon(:), 1, data(:));
colormap('Jet');
h=colorbar();
plotm(coast.coastlat,coast.coastlon,'k');
set(get(h, 'title'), 'string', units, ...
                  'Interpreter', 'none');
title({FILE_NAME; DATAFIELD_NAME}, 'Interpreter', 'none');
saveas(g, [FILE_NAME '.m.png']);
exit;

File: MORE/LAADS/MOD/MOD04_3K.A2022022.2355.061.2022024190628.hdf.py
"""

This example code illustrates how to regrid a LAADS MOD04_3K Swath
file in Python.

If you have any questions, suggestions, or comments on this example, please use
the HDF-EOS Forum (http://hdfeos.org/forums).  If you would like to see an
example of any other NASA HDF/HDF-EOS data product that is not listed in the
HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), feel free to
contact us at eoshelp@hdfgroup.org or post it at the HDF-EOS Forum
(http://hdfeos.org/forums).

Usage:  save this script and run

    $python MOD04_3K.A2022022.2355.061.2022024190628.hdf.py

The HDF file must be in your current working directory.

Tested under: Python 3.9.1 :: Miniconda
Last updated: 2022-01-27
"""
import os

import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
from mpl_toolkits.basemap import Basemap

from pyhdf.SD import SD, SDC

FILE_NAME = 'MOD04_3K.A2022022.2355.061.2022024190628.hdf'

# DATAFIELD_NAME = 'Angstrom_Exponent_1_Ocean'
DATAFIELD_NAME ='Optical_Depth_Land_And_Ocean'

# Open file.
hdf = SD(FILE_NAME, SDC.READ)

# Read dataset.
data2D = hdf.select(DATAFIELD_NAME)
data = data2D[:,:].astype(np.double)

# Read geolocation dataset.
lat = hdf.select('Latitude')
latitude = lat[:,:]
lon = hdf.select('Longitude')
longitude = lon[:,:]

# Retrieve attributes.
attrs = data2D.attributes(full=1)
aoa=attrs["add_offset"]
add_offset = aoa[0]
fva=attrs["_FillValue"]
_FillValue = fva[0]
sfa=attrs["scale_factor"]
scale_factor = sfa[0]        
ua=attrs["units"]
units = ua[0]
data[data == _FillValue] = np.nan
data = (data - add_offset) * scale_factor

# Draw plot.
m = Basemap(projection='cyl', resolution='l',
            llcrnrlat=-90, urcrnrlat = 90,
            llcrnrlon=-180, urcrnrlon = 180)
        
m.drawcoastlines(linewidth=0.5)
m.drawparallels(np.arange(-90., 120., 30.))
m.drawmeridians(np.arange(-180, 180., 45.))
m.scatter(lon, lat, c=data, s=1, cmap=plt.cm.jet,
          edgecolors=None, linewidth=0)
cb = m.colorbar()
cb.set_label(units)


# Put title using the first file.
basename = os.path.basename(FILE_NAME)
plt.title('{0}\n{1}'.format(basename, DATAFIELD_NAME))
fig = plt.gcf()

# Save image.
pngfile = "{0}.py.png".format(basename)
fig.savefig(pngfile)
    
